### PR TITLE
save_analysis: better handle paths and functions signature

### PR DESCRIPTION
**Pull request by marmeladema**
This should improve slightly some possible regressions due to hir rework.

r? @Xanewok

### CODE REVIEW

File: src/librustc_save_analysis/lib.rs [hunk 1]
```diff
@@ -144,7 +144,23 @@ impl<'l, 'tcx> SaveContext<'l, 'tcx> {
...
                     qualname,
-                    value: make_signature(decl, generics),
+                    value: fn_to_string(
+                        decl,
+                        hir::FnHeader {
+                            // functions in extern block are implicitly unsafe
+                            unsafety: hir::Unsafety::Unsafe,
+                            // functions in extern block cannot be const
+                            constness: hir::Constness::NotConst,
+                            abi: self.tcx.hir().get_foreign_abi(item.hir_id),
```
Comment (marmeladema) [inline src/librustc_save_analysis/lib.rs:154]: I managed to get the real abi of the extern block using `get_foreign_abi`. I don't think we should change the hir to store a `FnHeader` in `ForeignItem::Fn`. I tried it locally, it works, but its really invasive and somehow orthogonal to this change.

Comment (Xanewok) [reply src/librustc_save_analysis/lib.rs:154]: I think you're right, let's stick with what we have. Thanks!

**Review by Xanewok — APPROVED** [review]
Thanks!

**Comment by est31** [comment]
Thanks @marmeladema for fixing the bug! I've thought that with rls around the corner it would never be fixed.

**Comment by marmeladema** [comment]
I am actually using save-analysis issues as a way to learn more about the compiler :+1: 
save-analysis / rls might not stay around for long because rust-analyzer is kind of its successor i believe but in the mean time, if solving old issues can help some people then its not totally a lost cause i guess

**Pull request merged**